### PR TITLE
Initial Openapi3 support

### PR DESCRIPTION
**Pull request by opqdonut**
For #84

Replaces #563

This should be a minimum releasable version of openapi3 support now.

Missing bits which I'll address in future PRs:
- integrate more with muuntaja (get default content type etc)
- spec coercion doesn't generate `:additionalProperties false` unlike malli/schema
- multipart support

### CODE REVIEW

**Comment by opqdonut** [comment]
I think this is now ready for merge. The PR introduces new features, but doesn't break old ones. Openapi support is marked as alpha.

There's still a some refactoring to do, and some ease-of-use features (like proper muuntaja integration). I bet users will also discover bugs.

**Comment by ikitommi** [comment]
I'll look into this early next week.

File: modules/reitit-core/src/reitit/coercion.cljc [hunk 1]
```diff
@@ -98,17 +111,24 @@
...
+                                         [format (-response-coercer coercion schema)])
+                                       (filter second)
+                                       (seq)
+                                       (into {}))
+          default (when body (-response-coercer coercion body))]
+      (when (or per-format-coercers default)
+        (fn [request response]
+          (let [format (extract-response-format request response)
+                value (:body response)
+                coercer (get per-format-coercers format (or default (fn [value _format] value)))
```
Comment (ikitommi) [inline modules/reitit-core/src/reitit/coercion.cljc:127]: the fn `(fn [value _format] value)` is used in both request & response coercion and is generated per each request, could be extracted into a named fn in this ns.

File: modules/reitit-openapi/src/reitit/openapi.cljc [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+           accept-route (fn [route]
+                          (-> route second :openapi :id (or ::default) (trie/into-set) (set/intersection ids) seq))
+           ;base-openapi-spec {:responses ^:displace {:default {:description ""}}}
+           transform-endpoint (fn [[method {{:keys [coercion no-doc openapi] :as data} :data
+                                            middleware :middleware
+                                            interceptors :interceptors}]]
+                                (if (and data (not no-doc))
+                                  [method
+                                   (meta-merge
+                                    #_base-openapi-spec
```
Comment (ikitommi) [inline modules/reitit-openapi/src/reitit/openapi.cljc:100]: why is this commented out?

File: test/cljc/reitit/ring_coercion_test.cljc [hunk 3]
```diff
@@ -582,6 +584,79 @@
         (is (= {:status 200, :body {:total "FOO: this, BAR: that"}} (call m/schema custom-meta-merge-checking-schema)))
         (is (= {:status 200, :body {:total "FOO: this, BAR: that"}} (call identity custom-meta-merge-checking-parameters)))))))
 
+(deftest per-content-type-test
+  (doseq [[coercion json-request edn-request default-request json-response edn-response default-response]
+          [[#'malli/coercion
```
Comment (ikitommi) [inline test/cljc/reitit/ring_coercion_test.cljc:589]: why Vars here?

File: modules/reitit-spec/src/reitit/coercion/spec.cljc [hunk 4]
```diff
@@ -105,6 +107,40 @@
                               (if (:schema $)
                                 (update $ :schema #(coercion/-compile-model this % nil))
                                 $))]))})))
+        :openapi (openapi/openapi-spec
```
Comment (ikitommi) [inline modules/reitit-spec/src/reitit/coercion/spec.cljc:110]: could be extracted into a standalone function

File: modules/reitit-schema/src/reitit/coercion/schema.cljc [hunk 5]
```diff
@@ -67,6 +69,37 @@
                               (if (:schema $)
                                 (update $ :schema #(coercion/-compile-model this % nil))
                                 $))]))})))
+        :openapi (merge
```
Comment (ikitommi) [inline modules/reitit-schema/src/reitit/coercion/schema.cljc:72]: could be extracted into a standalone function

**Review by ikitommi — APPROVED** [review]
Looks Good! pushed a small commit to add openapi module to the build.

File: examples/http-swagger/src/example/server.clj [hunk 6]
```diff
@@ -42,11 +43,25 @@
...
         {:get {:no-doc true
                :swagger {:info {:title "my-api"
-                                :description "with reitit-http"}}
+                                :description "with reitit-http"}
+                         ;; used in /secure APIs below
+                         :securityDefinitions {"auth" {:type :apiKey
+                                                       :in :header
+                                                       :name "Example-Api-Key"}}}
                :handler (swagger/create-swagger-handler)}}]
+       ["/openapi.json"
```
Comment (ikitommi) [inline examples/http-swagger/src/example/server.clj:52]: we could change the default into swagger-ui handler to read the `openapi.json` over `swagger.json` if that exists?

**Comment by ikitommi** [comment]
more examples could be ported to use (just) the openapi in later documentation PRs?

**Comment by RickMoynihan** [comment]
I've been having a look at this PR too... Firstly thanks for putting it together, I'm very excited to see this is happening!

It seems that in order to get the swagger UI to work with the `http-swagger` example in the repo you also need to bump the swagger-ui dependency to be `[metosin/ring-swagger-ui "5.0.0-alpha.0"]` so it works with openapi `3.1.0`, it looks like @opqdonut has already done the bulk of the work by updating this dep upstream, but I think it may want updated here too.

**Comment by opqdonut** [comment]
@RickMoynihan Yeah, I'm depending on the alpha swagger-ui in the example right here:
https://github.com/metosin/reitit/pull/588/files#diff-110539d4d838e0f807261c996685cb4c5e32ee025a097f1c5541b816a6ebd6aaR7

**Comment by RickMoynihan** [comment]
ahh so you are!  Not sure how I missed that!  Apologies.

File: test/cljc/reitit/ring_coercion_test.cljc (hunk 3 above)
Comment (opqdonut) [reply test/cljc/reitit/ring_coercion_test.cljc:589]: Slightly nicer prints from the test:
```clj
user=> (str malli/coercion)
"reitit.coercion.malli$create$reify__9137@59568a15"
user=> (str #'malli/coercion)
"#'reitit.coercion.malli/coercion"
```

File: modules/reitit-openapi/src/reitit/openapi.cljc (hunk 2 above)
Comment (opqdonut) [reply modules/reitit-openapi/src/reitit/openapi.cljc:100]: no idea, left over from the previous PR #563. deleted.

File: modules/reitit-core/src/reitit/coercion.cljc (hunk 1 above)
Comment (opqdonut) [reply modules/reitit-core/src/reitit/coercion.cljc:127]: done

**Comment by opqdonut** [comment]
Fixed a couple of Tommi's comments. I intend to merge this latest tomorrow.

**Comment by RickMoynihan** [comment]
@opqdonut: I've noticed a small problem with the `http-swagger` example which I thought I'd report here.

If you change the `:url` for the swagger UI by passing `:url "/openapi.json"` to `swagger-ui/create-swagger-ui-handler`, run the app and try to hit the route `/files/upload` you get a 400 bad request with the following spec error:

```
{
  "spec": "(spec-tools.core/spec {:spec (clojure.spec.alpha/keys :req-un [:spec$28049/file]), :type :map, :leaf? false})",
  "problems": [
    {
      "path": [],
      "pred": "(clojure.core/fn [%] (clojure.core/contains? % :file))",
      "val": {},
      "via": [],
      "in": []
    }
  ],
  "type": "reitit.coercion/request-coercion",
  "coercion": "spec",
  "value": {},
  "in": [
    "request",
    "multipart-params"
  ]
}
```

Additionally, the swagger UI doesn't render the nice upload button that you get if you set the url to `/swagger.json`; I can imagine this later one is an upstream issue with the UI, but the former looks to be something to do with the multipart interceptor.

**Comment by kjothen** [comment]
I think the example comment is wrong...it should read `create-openapi-handler`, not `openapi-spec-handler`.

``` clojure
       [\"/openapi.json\"
        {:get {:no-doc true
               :openapi {:info {:title \"my-api\"}}
               :handler reitit.openapi/openapi-spec-handler}}]
```

**Comment by ikitommi** [comment]
guide on OAS3 multipart here: https://swagger.io/docs/specification/describing-request-body/multipart-requests/

**Comment by ikitommi** [comment]
small commit to configure the example swagger-ui to present both swagger & openapi definitions.

Screenshot 2023-03-14 at 21 03 06

**Comment by opqdonut** [comment]
I'll merge this now since it's a useful baseline. I'll fix multipart  support in a follow-up PR.

File: examples/http-swagger/src/example/server.clj [hunk 7]
```diff
@@ -52,7 +52,7 @@
        ["/openapi.json"
         {:get {:no-doc true
                :openapi {:info {:title "my-api"
-                                :description "with reitit-http"
+                                :description "openap-docs with reitit-http"
```
Comment (RickMoynihan) [inline examples/http-swagger/src/example/server.clj:56]: I know this is merged, but there's a small typo here:

Should be `openapi-docs ...`

File: examples/http-swagger/src/example/server.clj [hunk 8]
```diff
@@ -142,6 +175,9 @@
       (swagger-ui/create-swagger-ui-handler
         {:path "/"
          :config {:validatorUrl nil
+         	        :urls [{:name "swagger", :url "swagger.json"}
```
Comment (Deraen) [inline examples/http-swagger/src/example/server.clj:178]: Bad indent / tabs instead of spaces here.

Comment (opqdonut) [reply examples/http-swagger/src/example/server.clj:178]: in #592

File: examples/http-swagger/src/example/server.clj (hunk 7 above)
Comment (opqdonut) [reply examples/http-swagger/src/example/server.clj:56]: in #592

**Pull request merged**